sql/catalyst/src/main/java/org/apache/spark/sql/RowFactory.java

```diff
@@ -37,4 +39,17 @@
   public static Row create(Object ... values) {
     return new GenericRow(values);
   }
+
+  /**
+   * Create a {@link Row} from the given arguments. Provided schema is incorporated into
+   * created {@link Row} object, and allows getAs(fieldName) to access the value of column.
+   *
+   * Note that every Rows will contain the duplicated schema, hence in high volume it is still
+   * recommended to use `create` with accessing column by position.
+   *
+   * @since 3.0.0
+   */
+  public static Row createWithSchema(StructType schema, Object ... values) {
```

Comment:
This change looks useful!

Reply:
This is probably OK, but what's the use case?
Nit: remove space before `...` and can you use the `@Since` annotation?
Any test at all for it would be great too.

Reply:
This forces Row to have schema inside which is necessary to use `getAs(fieldName)` in downstream operator.

Please take a look at the query I've added in PR description. You can see Spark provides `GenericRowWithSchema` in MapFunction, but when users create Row with `RowFactory.
create`, newly created rows don't have schema and not able to use `getAs(fieldName)`. 

More weird thing is, the schema information is conditionally provided, depending on the query plan. (adding withWatermark in prior to filter makes getAs(fieldName) be available even we return `RowFactory.create` in map.

But in overall I agree with @HyukjinKwon that fixing root cause would be ideal. This is just a way to mitigation, and I wouldn't mind the mitigation would not be accepted. If you feel OK with taking workaround I'll address your comment. Thanks!

Btw, actually I feel this sounds the matter of interface contract - Row interface should provide getAs(fieldName) regardless of its implementation (otherwise methods should have doc that getAs(fieldName) is depending on implementation and implementation may just throw exception), and if not, Row should remove the methods and some other interface should extend Row and have them.